service/index/writer.go

```diff
@@ -150,3 +169,40 @@ func (w *Writer) Events(height uint64, events []flow.Event) error {
 	}
 	return nil
 }
+
+func (w *Writer) apply(op func(*badger.Txn) error) error {
+	w.RLock()
+	err := w.err
+	w.RUnlock()
+	if err != nil {
+		return fmt.Errorf("could not commit transaction: %w", w.err)
+	}
+	err = op(w.tx)
+	if errors.Is(err, badger.ErrTxnTooBig) {
+		w.tx.CommitWith(w.done)
+		_ = w.sema.Acquire(context.Background(), 1)
+		w.tx = w.db.NewTransaction(true)
+		err = op(w.tx)
+	}
+	if err != nil {
+		return fmt.Errorf("could not apply operation: %w", err)
+	}
+	return nil
+}
+
+func (w *Writer) done(err error) {
+	if err != nil {
+		w.Lock()
+		w.err = err
+		w.Unlock()
+	}
+	w.sema.Release(1)
+}
+
+func (w *Writer) Close() error {
+	_ = w.sema.Acquire(context.Background(), int64(w.cfg.ConcurrentTransactions))
+	if w.err != nil {
+		return fmt.Errorf("could not flush transactions: %w", w.err)
+	}
```

Comment:
Maybe I am missing something but it seems like we might panic by accessing `w.err` here without waiting for the lock to be available.

Reply:
I think it's fair to assume we are no longer applying new updates when we close the writer. This means that, at most, some calls are still waiting on the semaphore. As we acquire all the available slots from the semaphore, all transactions will have finished, and there will be no more `done` callbacks.